solr/solrj/src/java/org/apache/solr/common/cloud/NodesSysPropsCacher.java

```diff
@@ -17,193 +17,87 @@
-            }
-          }
-          cache.put(node, Collections.unmodifiableMap(props));
-        } else {
-          fetchRemoteProps(node, fetchRound);
-        }
-      }
-    }
-  }
-
-  private void fetchRemoteProps(String node, int fetchRound) {
-    for (int i = 0; i < NUM_RETRY; i++) {
-      if (isClosed && fetchRound != fetchCounting.get()) return;
-
-      try {
-        Map<String, Object> props = nodeStateProvider.getNodeValues(node, tags);
-        cache.put(node, Collections.unmodifiableMap(props));
-        break;
-      } catch (Exception e) {
-        try {
-          // 1, 4, 9
-          int backOffTime = 1000 * (i + 1);
-          backOffTime = backOffTime * backOffTime;
-          backOffTime = Math.min(10000, backOffTime);
-          Thread.sleep(backOffTime);
-        } catch (InterruptedException e1) {
-          Thread.currentThread().interrupt();
-          log.info(
-              "Exception on caching node:{} system.properties:{}, retry {}/{}",
-              node,
-              tags,
-              i + 1,
-              NUM_RETRY,
-              e); // nowarn
-          break;
-        }
-        log.info(
-            "Exception on caching node:{} system.properties:{}, retry {}/{}",
-            node,
-            tags,
-            i + 1,
-            NUM_RETRY,
-            e); // nowarn
-      }
-    }
-  }
-
-  public Map<String, Object> getSysProps(String node, Collection<String> tags) {
-    Map<String, Object> props = cache.get(node);
-    HashMap<String, Object> result = new HashMap<>();
-    if (props != null) {
-      for (String tag : tags) {
-        if (props.containsKey(tag)) {
-          result.put(tag, props.get(tag));
-        }
+  public Map<String, Object> getSysProps(String nodeName, Collection<String> tags) {
+    Map<String, Object> cached =
+        nodeVsTagsCache.computeIfAbsent(nodeName, s -> new LinkedHashMap<>());
+    Map<String, Object> result = new LinkedHashMap<>();
+    for (String tag : tags) {
+      if (!cached.containsKey(tag)) {
+        // at least one property is missing. fetch properties from the node
+        Map<String, Object> props = fetchProps(nodeName, tags);
+        // make a copy
+        cached = new LinkedHashMap<>(cached);
+        // merge all properties
+        cached.putAll(props);
+        // update the cache with the new set of properties
+        nodeVsTagsCache.put(nodeName, cached);
+        return props;
+      } else {
+        result.put(tag, cached.get(tag));
       }
     }
     return result;
   }
 
-  public int getCacheSize() {
-    return cache.size();
-  }
-
-  public boolean isRunning() {
-    return isRunning.get();
-  }
+  private Map<String, Object> fetchProps(String nodeName, Collection<String> tags) {
+    SolrParams p = new ModifiableSolrParams();
+    StringBuilder sb = new StringBuilder(zkStateReader.getBaseUrlForNodeName(nodeName));
+    sb.append("/admin/metrics?omitHeader=true&wt=javabin");
```

Comment:
Shouldn't we be using Solr's standard request/response APIs with Params etc.?

Reply:
There is no clean way to use Solr's standard client to send request and parse response from a single server directly

Reply:
Can you just create an Http2SolrClient with that URL?

Reply:
I don't need a SolrClient. I just need an HTTP client. The Solr client APIs are pretty clumsy

Reply:
Agreed SolrClient should be simpler for this case, although I don't think it's bad.  It at least handles URL encoding/decoding whereas using this HTTP client does not.  Its debatable which is better.  In our codebase, SolrClient is used basically everywhere and not a plain HTTP client so as a reviewer I'm wondering why here?  Maybe there are infrastructural details taken care of here like distributed tracing, authentication/authorization, and SSL with SolrClient that are not with a plain HTTP client?  If you can rationalize that these don't matter or will work then okay.

Reply:
On the other hands, if we're using HTTP client because SolrClient can't handle these endpoints, then let's fix Solr Client to be better instead of trying to band aid around it

Reply:
@madrob , I would love to do that

Reply:
I have modified the http call invocation to use a `GenericSolrRequest`